webview-ui/src/utils/transformation.ts

```diff
@@ -31,5 +31,25 @@ export const groupIncidentsByMsg = (
         acc[msg].push([uri, incident]);
         return acc;
       },
-      {} as { [msg: string]: [string, Incident][] },
+      {} as { [msg: string]: [string, EnhancedIncident][] },
     );
+
+export function enhanceIncidents(
```

Comment:
I wonder, if instead of having this function, we updated the loadRulesets to return an `EnhancedIncident[]`. At that time we have all of the information we need to properly construct these incidents. If we pull that thread to completion, we can eliminate the need for dragging rulesets and violations all around our codebase.